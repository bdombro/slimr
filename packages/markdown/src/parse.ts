/**
 * Converts a limited markdown subset to HTML
 *
 * - headings, i.e. # H1, ## H2, ### H3, #### H4, ##### H5, ###### H6
 * - alt H1 heading, i.e. H1\n===
 * - alt H2 heading, i.e. H1\n---
 * - Horizontal rule/lines, i.e. *** --- ___
 * - **bold**
 * - *italic*
 * - ~~strikethrough~~
 * - links, i.e. [Link](https://google.com)  [Link with title](https://google.com 'title')
 * - reference links, i.e. [link1-text][link1-ref]
 * - images, i.e. ![Image](https://google.com)  ![Image with title](https://google.com 'title')
 * - ordered and unordered lists up to one level. nesting not supported.
 * - code blocks
 * - Most HTML is passed through without modification, except scripts/style which will be HTML encoded
 *
 * Common alternatives: snarkdown, markdown-it, marked
 *
 * @param md
 * A string containing valid markdown text
 *
 * @returns
 * A string of html
 *
 * @example
 * ```typescript
 * import { parse } from '@slimr/markdown'
 * const html = parse(`
 *  # ~~The Jungle~~Heaven
 *
 *  Welcome to ~~the jungle~~heaven, baby.
 * `)
 * ```
 */
export const parse = (md: string) => {
  const trashgc: string[] = []

  md = md.trim().replace(/\n[ \t]+/gm, '\n') // trim indentation

  if (!md) return ''

  md = md.replace(/<(\/)?(script|style)>/gim, '&lt;$1$2&gt;') // Encode any script and style tags

  // an array to store the intermediate placeholders
  // Create keys and add chunks to it that we don't want mangle
  const placeholders: [key: string, value: string][] = []

  // code blocks i.e. ```code```
  // transpile and remove before others to prevent interference
  md = md
    .replace(/```/g, 'ỻ') // intermediate placeholder
    .replace(/ỻ([^ỻ\n]*)ỻ/gm, (_, code) => `<code>${code}</code>`)
    .replace(/ỻ(.*)(\n*)([^ỻ]*)ỻ/gm, (_, ...args) => {
      let code = `<code${args[0] ? ` class="language-${args[0]}"` : ''} style="padding-top:0">${
        args[1] + args[2].replace(/&/g, '&amp').replace(/</g, '&lt').replace(/>/g, '&gt')
      }</code>`
      if (code.includes('\n')) {
        code = `<pre>${code}</pre>`
      }
      const key = 'ỻ' + placeholders.length
      placeholders.push([key, code])
      return key
    })

  md = md
    .replace(/\n\n(\*\*\*+|---*|___*)$/gm, '\n<hr/>') // Horizontal rule/lines, i.e. *** --- ___
    .replace(/^###### (.*$)(\n*)/gm, '<h6>$1</h6>\n')
    .replace(/^##### (.*$)(\n*)/gm, '<h5>$1</h5>\n')
    .replace(/^#### (.*$)(\n*)/gm, '<h4>$1</h4>\n')
    .replace(/^### (.*$)(\n*)/gm, '<h3>$1</h3>\n')
    .replace(/^## (.*$)(\n*)/gm, '<h2>$1</h2>\n')
    .replace(/^# (.*$)(\n*)/gm, '<h1>$1</h1>\n')
    .replace(/^(.*$)\n===+\n\n*/gm, '<h1>$1</h1>\n') // alt h1
    .replace(/^(.*$)\n---+\n\n*/gm, '<h2>$1</h2>\n') // alt h2
    .replace(/^> (.*$)/gm, '<blockquote>$1</blockquote>') // blockquotes i.e. '> quote'
    // if a line ends with a space, replace with &nbsp;
    .replace(/ $/gm, '&nbsp;')
    .replace(/\*\*(.*?)\*\*/gm, '<b>$1</b>') // bold
    .replace(/\*(.*?)\*/gm, '<em>$1</em>') // italic
    .replace(/~~(.*?)~~/gm, '<strike>$1</strike>') // deleted text
    // images
    .replace(/!\[(.*?)\]\(([^ ]*)\)/gm, "<img alt='$1' src='$2'/>")
    .replace(
      /!\[(.*?)\]\(([^ ]*)( [('"][^)'"]+[)'"])*\)/gm,
      (_, ...args) =>
        `<img alt='${args[0]}' src='${args[1]}'${
          args[2] ? ` title='${args[2].slice(2, -1)}'` : ''
        }/>`
    )
    // links
    .replace(
      /\[(.*?)\]\(([^ ]*)( [('"][^)'"]+[)'"])*\)/gm,
      (_, ...args) =>
        `<a href='${args[1]}'${args[2] ? ` title='${args[2].slice(2, -1)}'` : ''}>${args[0]}</a>`
    )
    .replace(/([^ \n]+@[^. \n]+\.[^ \n]+)/gm, "<a href='mailto:$1'>$1</a>") // emails
    // reference links
    .replace(/\[([^\]]+)\]\[([^\]]+)\]/gm, (_, ...args) => {
      const refValue = new RegExp('\\[' + args[1] + '\\]: ([^ \n]+)( [^\n]*)?\n', 'gim').exec(md)
      if (refValue !== null) {
        trashgc.push(refValue[0])
        return (
          '<a href="' +
          refValue[1] +
          '"' +
          (refValue[2] ? ` title=${refValue[2].trim().replace(/[()]/g, '"')}` : '') +
          '>' +
          args[0] +
          '</a>'
        )
      } else return _
    })
    // unordered lists i.e. - listitem
    .replace(/\n\n+(- (.*))$/gm, '\n<ul>\n$1')
    .replace(/^(- (.*))\n\n/gm, '$1\n</ul>\n\n')
    .replace(/^- (.*)$/gm, '<li>$1</li>')
    // ordered lists i.e. 1. listitem
    .replace(/\n\n+(\d\. (.*))$/gm, '\n<ol>\n$1')
    .replace(/^(\d\. (.*))\n\n/gm, '$1\n</ol>\n\n')
    .replace(/^\d\. (.*)$/gm, '<li>$1</li>')

  // Trash collect
  for (const s of trashgc) {
    md = md.replace(s, '')
  }

  if (!md.endsWith('\n')) {
    md += '\n'
  }

  md = md
    .replace(/([\w '"])\n([\w '"])/gm, '$1 $2') // replace newlines between words with a space
    .replace(/^((\w|'|"|&|<b>|<s>|<a>|<a\s|<strong>|<strike>|<i>)[^\n]+)\n$/gm, '<p>$1</p>')

  // Add placeholders back
  for (const [key, value] of placeholders) {
    md = md.replace(key, value)
  }

  md = '\n' + md.trim() + '\n'

  return md
}
