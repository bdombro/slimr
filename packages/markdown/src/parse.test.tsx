import {parse} from './parse.js'

it('parses correctly', () => {
  const markdown = `
# H1
## H2
### H3
#### H4
##### H5
###### H6
this text should not interfere with the above


Alt-H1
=============

Alt-H2
-------------
this text should not interfere with the above

soft
return

**Bold** *Italic* ~~Strikethrough~~

[Link](https://google.com)  [Link with title](https://google.com 'title') [@not-an-email](http://google.com)

![Image](https://google.com)  ![Image with title](https://google.com 'title')

<img src="https://google.com" alt="alt" title="title" />

<div>
  <p>
    Most html is left alone.
  </p>
</div>

email@email.com notemail@emailcom email@email.com


*****

*****

-----

_____
this text should not interfere with the above

> Blockquote

\`\`\`
// unknown
multi
\`line\`
generic code
\`\`\`

\`\`\`bash
# bash
multi
\`line\`
bash code
\`\`\`

Here is \`\`\`Inline Code\`\`\`.

Reference links [link1-text][link1] [link2-text][link2] [link3-text][link3]

[link1]: /link1
[link2]: /link2 'title2'
[link3]: /link3 (title3)

- li1
- li2
- li3

1. li4
2. li5
3. li6

<script>This script tag is HTML encoded for security</script>

<style>This style tag is HTML encoded for security</style>
`
  const parsed = parse(markdown)
  // console.log(parsed)
  expect(parsed).toMatchSnapshot()
})
