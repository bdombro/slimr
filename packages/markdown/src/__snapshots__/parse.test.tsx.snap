// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parses correctly 1`] = `
"<h1>H1</h1>
<h2>H2</h2>
<h3>H3</h3>
<h4>H4</h4>
<h5>H5</h5>
<h6>H6</h6>
<p>this text should not interfere with the above</p>

<h1>Alt-H1</h1>
<h2>Alt-H2</h2>
<p>this text should not interfere with the above</p>
<p>soft return</p>
<p><b>Bold</b> <i>Italic</i> <strike>Strikethrough</strike></p>
<p><a href='https://google.com'>Link</a>  <a href='https://google.com' title='title'>Link with title</a> <a href='http://google.com'>@not-an-email</a></p>
<img alt='Image' src='https://google.com'/>  <img alt='Image with title' src='https://google.com' title='title'/>

<img src="https://google.com" alt="alt" title="title" />

<div>
<p>
Most html is left alone.
</p>
</div>

<p><a href='mailto:email@email.com'>email@email.com</a> notemail@emailcom <a href='mailto:email@email.com'>email@email.com</a></p>
<hr/>
<hr/>
<hr/>
<hr/>
<p>this text should not interfere with the above</p>
<blockquote>Blockquote</blockquote>

<pre><code style="padding-top:0">
// unknown
multi
\`line\`
generic code
</code></pre>

<pre><code class="language-bash" style="padding-top:0">
# bash
multi
\`line\`
bash code
</code></pre>

<p>Here is <code>Inline Code</code>.</p>
<p>Reference links <a href="/link1">link1-text</a> <a href="/link2" title='title2'>link2-text</a> <a href="/link3" title="title3">link3-text</a></p>
<ul>
<li>li1</li>
<li>li2</li>
<li>li3</li>
</ul>
<ol>
<li>li4</li>
<li>li5</li>
<li>li6</li>
</ol>

<p>&lt;script&gt;This script tag is HTML encoded for security&lt;/script&gt;</p>
<p>&lt;style&gt;This style tag is HTML encoded for security&lt;/style&gt;</p>"
`;
